.circleci/config.yml

```diff
@@ -41,18 +41,20 @@ jobs:
       - checkout
       - run: git submodule sync
       - run: git submodule update --init
+      - run: git fetch origin main  # needed for comparisson in pre-commit
+      - run: git branch --track main origin/main  # needed for comparisson in pre-commit
       - run: pip install --user tox
```

Comment:
Shouldn't we pin this too?

Reply:
It's not currently monitored by anything so I wanted to "cope" with the failures by fixing them. We won't get alerts when there are upstream changes. We could consider `tox<5` to be fairly low-risk.

Reply:
Pinning is also a good idea because all currently-open PRs now have to merge these fixes.

Reply:
Yea, I'm not sure why we weren't pinning this before, but it's definitely a best practice. So we should pin each of them.

Reply:
Added the upper bound :+1: